File: tools/ghidra/ExportFunctions.java
// Ghidra Java post-analysis script
// Exports function list with addresses, sizes, and names
//
// @category NDS
// @author MLPiT Decomp

import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.*;
import ghidra.program.model.address.*;
import java.io.*;
import java.util.*;

public class ExportFunctions extends GhidraScript {

    @Override
    public void run() throws Exception {
        FunctionManager fm = currentProgram.getFunctionManager();

        String outputDir = System.getenv("MLPIT_OUTPUT_DIR");
        if (outputDir == null || outputDir.isEmpty()) {
            outputDir = ".";
        }

        // Collect all functions
        List<String[]> functions = new ArrayList<>();
        FunctionIterator iter = fm.getFunctions(true);
        int thunkCount = 0;
        long totalCodeBytes = 0;

        while (iter.hasNext()) {
            Function f = iter.next();
            AddressSetView body = f.getBody();
            long size = body.getNumAddresses();
            totalCodeBytes += size;

            if (f.isThunk()) thunkCount++;

            functions.add(new String[]{
                f.getEntryPoint().toString(),
                String.valueOf(size),
                f.getName(),
                String.valueOf(f.isThunk()),
                String.valueOf(f.getParameterCount())
            });
        }

        // Write human-readable function map
        String txtPath = outputDir + File.separator + "functions.txt";
        PrintWriter txtWriter = new PrintWriter(new FileWriter(txtPath));
        txtWriter.println("# Auto-generated function map from Ghidra analysis");
        txtWriter.println("# Address        Size  Name");
        txtWriter.println("#" + "-".repeat(70));
        for (String[] func : functions) {
            txtWriter.printf("  %-14s %6s  %s%n", func[0], func[1], func[2]);
        }
        txtWriter.printf("%n# Total functions: %d%n", functions.size());
        txtWriter.close();

        // Write CSV for easy processing
        String csvPath = outputDir + File.separator + "functions.csv";
        PrintWriter csvWriter = new PrintWriter(new FileWriter(csvPath));
        csvWriter.println("address,size,name,is_thunk,param_count");
        for (String[] func : functions) {
            csvWriter.printf("%s,%s,%s,%s,%s%n", func[0], func[1], func[2], func[3], func[4]);
        }
        csvWriter.close();

        // Print summary
        println("");
        println("==================================================");
        println("Analysis Export Summary");
        println("==================================================");
        println("  Functions found:    " + functions.size());
        println("  Total code bytes:   " + totalCodeBytes);
        println("  Thunk functions:    " + thunkCount);
        println("  Output (txt):       " + txtPath);
        println("  Output (csv):       " + csvPath);
        println("==================================================");
    }
}

File: tools/ghidra/export_functions.py
# Ghidra headless post-analysis script
# Exports function list, entry points, and basic stats
#
# @category NDS
# @author MLPiT Decomp

import json
import os

def run():
    program = currentProgram
    listing = program.getListing()
    fm = program.getFunctionManager()
    mem = program.getMemory()

    output_dir = os.environ.get("MLPIT_OUTPUT_DIR", ".")

    # Collect all functions
    functions = []
    func_iter = fm.getFunctions(True)
    while func_iter.hasNext():
        f = func_iter.next()
        body = f.getBody()
        size = 0
        for rng in body:
            size += rng.getLength()
        functions.append({
            "name": f.getName(),
            "address": str(f.getEntryPoint()),
            "size": size,
            "is_thunk": f.isThunk(),
            "calling_convention": str(f.getCallingConventionName()),
            "param_count": f.getParameterCount(),
        })

    # Sort by address
    functions.sort(key=lambda x: x["address"])

    # Write function list as JSON
    json_path = os.path.join(output_dir, "functions.json")
    with open(json_path, "w") as fp:
        json.dump(functions, fp, indent=2)

    # Write human-readable function map
    txt_path = os.path.join(output_dir, "functions.txt")
    with open(txt_path, "w") as fp:
        fp.write("# Auto-generated function map\n")
        fp.write("# Address      Size  Name\n")
        fp.write("#" + "-" * 70 + "\n")
        for f in functions:
            fp.write("  %-12s %6d  %s\n" % (f["address"], f["size"], f["name"]))
        fp.write("\n# Total functions: %d\n" % len(functions))

    # Summary stats
    total_code_bytes = sum(f["size"] for f in functions)
    print("")
    print("=" * 50)
    print("Analysis Export Summary")
    print("=" * 50)
    print("  Functions found:    %d" % len(functions))
    print("  Total code bytes:   %d" % total_code_bytes)
    print("  Thunk functions:    %d" % sum(1 for f in functions if f["is_thunk"]))
    print("  Output:             %s" % json_path)
    print("  Output:             %s" % txt_path)
    print("=" * 50)

run()
